plugins/common/view-action-interface.cpp

```diff
@@ -0,0 +1,308 @@
+    }
+    else
+    {
+        auto arg_double = _expect_double(args, 1);
+        if (std::get<0>(arg_double)) {
+            return {true, static_cast<float>(std::get<1>(arg_double))};
+        }
+    }
+
+    std::cerr << "View action interface: Invalid arguments. Expected 'set alpha [float|double]." << std::endl;
+    return {false, 1.0f};
+}
+
+std::tuple<bool, int, int, int, int> view_action_interface_t::_validate_geometry(const std::vector<variant_t> &args)
+{
+    auto arg_x = _expect_int(args, 1);
+    auto arg_y = _expect_int(args, 2);
+    auto arg_w = _expect_int(args, 3);
+    auto arg_h = _expect_int(args, 4);
+
+    if (std::get<0>(arg_x) && std::get<0>(arg_y) && std::get<0>(arg_w) && std::get<0>(arg_h)) {
+        return {true, std::get<1>(arg_x), std::get<1>(arg_y), std::get<1>(arg_w), std::get<1>(arg_h)};
+    }
+
+    std::cerr << "View action interface: Invalid arguments. Expected 'set geometry int int int int." << std::endl;
+    return {false, 0, 0, 0, 0};
+}
+
+std::tuple<bool, int, int> view_action_interface_t::_validate_position(const std::vector<variant_t> &args)
+{
+    auto arg_x = _expect_int(args, 0);
+    auto arg_y = _expect_int(args, 1);
+
+    if (std::get<0>(arg_x) && std::get<0>(arg_y)) {
+        return {true, std::get<1>(arg_x), std::get<1>(arg_y)};
+    }
+
+    std::cerr << "View action interface: Invalid arguments. Expected 'move int int." << std::endl;
+    return {false, 0, 0};
+}
+
+std::tuple<bool, int, int> view_action_interface_t::_validate_size(const std::vector<variant_t> &args)
+{
+    auto arg_w = _expect_int(args, 0);
+    auto arg_h = _expect_int(args, 1);
+
+    if (std::get<0>(arg_w) && std::get<0>(arg_h)) {
+        return {true, std::get<1>(arg_w), std::get<1>(arg_h)};
+    }
+
+    std::cerr << "View action interface: Invalid arguments. Expected 'resize int int." << std::endl;
+    return {false, 0, 0};
+}
+
+void view_action_interface_t::_set_alpha(float alpha)
+{
+    alpha = std::clamp(alpha, 0.1f, 1.0f);
+
+    // Apply view transformer if needed and set alpha.
+    wf::view_2D *transformer;
+
+    if (!_view->get_transformer("alpha")) {
+        _view->add_transformer(std::make_unique<wf::view_2D> (_view), "alpha");
+    }
+
+    transformer = dynamic_cast<wf::view_2D *> (_view->get_transformer("alpha").get());
+    if (fabs(transformer->alpha - alpha) > FLT_EPSILON)
+    {
+        transformer->alpha = alpha;
+        _view->damage();
+
+        std::cout << "View action interface: Alpha set to " << alpha << "." << std::endl;
+    }
+}
+
+void view_action_interface_t::_set_geometry(int x, int y, int w, int h)
+{
+    _resize(w, h);
+    _move(x, y);
+}
+
+void view_action_interface_t::_move(int x, int y)
+{
+    // Clamp x and y to sane values. Do not allow to move outside of the output.
+    auto output = _view->get_output();
+    if (output != nullptr)
+    {
+        auto dimensions = output->get_screen_size();
+
+        x = std::clamp(x, 0, (dimensions.width - 40));
```

Comment:
Any rationale for this limit?

Reply:
I wanted to avoid moving the window to a position offscreen where it would be impossible to grab the title bar and manually move it back. To avoid "losing" the window due to a bad rule. I used the 40/30 pixels again as minimum size in resize method below.

Reply:
Hmm, I see. However, moving to negative coordinates or something else also has its use: for example, if I move to negative coordinates, I will be moving my view to the workspace on the left. Also, maybe I want to put part of the view offscreen.

All in all, here is what I think can fulfill both usecases: just reject moving the view iff it is not visible on any workspace. For example using this function: https://github.com/WayfireWM/wayfire/blob/master/src/api/wayfire/workspace-manager.hpp#L115

Reply:
I did not know about the workspaces thing. I will change the code to check if the move would go to another workspace and reject it if not.

Are you good with the minimum size limit I put on, or is there a trick that I missed there as well?

Reply:
No, there is no other trick (which is why I didn't comment on the resize code :)